massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,731 @@
+use massa_signature::{MassaSignatureError, PublicKey, Signature};
+
+//
+
+/// Can we build a Denunciation from this object?
+pub trait Denounceable {
+    /// Return true if object can be denounced, false otherwise
+    fn is_denounceable() -> bool {
+        false
+    }
+
+    /// Return DenunciationData (used to modify Signature for `SecureShare<T>`)
+    /// This is a default impl, return None if Self::is_denounceable() returns false
+    fn get_denunciation_data(&self) -> Option<DenunciationData> {
+        match Self::is_denounceable() {
+            true => unimplemented!(),
+            false => None,
+        }
+    }
+}
+
+/// Denunciation data (to be include in `SecureShare<T>` signature
+pub enum DenunciationData {
+    ///
+    Endorsement((Slot, u32)),
+    ///
+    BlockHeader(Slot),
+}
+
+/// A Serializer for `DenunciationData`
+#[derive(Default)]
+pub struct DenunciationDataSerializer {
+    slot_serializer: SlotSerializer,
+    index_serializer: U32VarIntSerializer,
+}
+
+impl DenunciationDataSerializer {
+    /// Create a new `DenunciationDataSerializer`
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Serializer<DenunciationData> for DenunciationDataSerializer {
+    fn serialize(
+        &self,
+        value: &DenunciationData,
+        buffer: &mut Vec<u8>,
+    ) -> Result<(), SerializeError> {
+        match value {
+            DenunciationData::Endorsement(data) => {
+                self.slot_serializer.serialize(&data.0, buffer)?;
+                self.index_serializer.serialize(&data.1, buffer)?;
+            }
+            DenunciationData::BlockHeader(data) => {
+                self.slot_serializer.serialize(data, buffer)?;
+            }
+        };
+        Ok(())
+    }
+}
+
+// Denounceable trait impl
+
+impl Denounceable for Endorsement {
+    fn is_denounceable() -> bool {
+        true
+    }
+    fn get_denunciation_data(&self) -> Option<DenunciationData> {
+        Some(DenunciationData::Endorsement((self.slot, self.index)))
+    }
+}
+
+impl Denounceable for BlockHeader {
+    fn is_denounceable() -> bool {
+        true
+    }
+    fn get_denunciation_data(&self) -> Option<DenunciationData> {
+        Some(DenunciationData::BlockHeader(self.slot))
+    }
+}
+
+impl Denounceable for Block {}
+impl Denounceable for Operation {}
+
+//
+
+#[allow(dead_code)]
+struct EndorsementDenunciation {
+    public_key: PublicKey,
+    slot: Slot,
+    index: u32,
+    hash_1: Hash,
+    hash_2: Hash,
+    signature_1: Signature,
+    signature_2: Signature,
+}
+
+impl EndorsementDenunciation {
+    /// Compute hash for Endorsement content
+    fn compute_content_hash(content: &Endorsement) -> Result<Hash, SerializeError> {
+        let mut buf = Vec::new();
+        let endorsement_serializer = EndorsementSerializer::new();
+        endorsement_serializer.serialize(content, &mut buf)?;
+        Ok(Hash::compute_from(&buf))
+    }
+
```

Comment:
Currently we avoid computing the hash twice and being sensitive to signature malleability by computing `hash(contents without signature)` to get the ID of the object, then this hash is the one directly signed/verified.

We should try to keep this behavior and simply redefine the ID computation as: `hash(denunciation_field1 + denunciation_field2 + hash(contents without signature))`.